### PR TITLE
Use Highway with Static Dispatch for some types in Absolute

### PR DESCRIPTION
**Pull request by Mousius**
This variant on #24384 uses the static dispatch within Highway, but the dynamic dispatch exists in NumPy. 

Similarly, loads and stores can only be contiguous rather than supporting all potential layouts (waiting for the Scatter equivalent for https://github.com/google/highway/commit/c6c09c402da635b3e0f2ea1a3a8c72acf91ad06f, then it should be trivial).

### CODE REVIEW

File: numpy/core/src/umath/absolute.dispatch.cpp [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+#define _UMATHMODULE
+#define _MULTIARRAYMODULE
+#define NPY_NO_DEPRECATED_API NPY_API_VERSION
+
+#define PY_SSIZE_T_CLEAN
+#include <Python.h>
+
+#undef HWY_TARGET_INCLUDE
+#define HWY_TARGET_INCLUDE "absolute.dispatch.cpp"  // this file
+#include <hwy/foreach_target.h>  // must come before highway.h
```
Comment (jan-wassenberg) [inline numpy/core/src/umath/absolute.dispatch.cpp:14]: This and the HWY_TARGET_INCLUDE can/should be commented out for static dispatch. We only want to compile the file once for one target.

File: numpy/core/src/umath/absolute.dispatch.cpp [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+// Alternative to per-function HWY_ATTR: see HWY_BEFORE_NAMESPACE
+template <typename T>
+HWY_ATTR void SuperAbsolute(char **args, npy_intp const *dimensions, npy_intp const *steps) {
+  const T* HWY_RESTRICT input_array = (const T*) args[0];
+  T* HWY_RESTRICT output_array = (T*) args[1];
+  const size_t size = dimensions[0];
+  const hn::ScalableTag<T> d;
+  
+  for (size_t i = 0; i < size; i += hn::Lanes(d)) {
+    const auto in = hn::Load(d, input_array + i);
```
Comment (jan-wassenberg) [inline numpy/core/src/umath/absolute.dispatch.cpp:37]: Are we sure the memory is aligned to the vector size? If not, we should use LoadU/StoreU.

File: numpy/core/src/umath/absolute.dispatch.cpp [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+HWY_EXPORT(INT_SuperAbsolute);
+HWY_EXPORT(FLOAT_SuperAbsolute);
+HWY_EXPORT(DOUBLE_SuperAbsolute);
+
+extern "C" {
+
+NPY_NO_EXPORT void
+INT_absolute(char **args, npy_intp const *dimensions, npy_intp const *steps, void *NPY_UNUSED(func))
+{
+  static auto dispatcher = HWY_STATIC_DISPATCH(INT_SuperAbsolute);
```
Comment (jan-wassenberg) [inline numpy/core/src/umath/absolute.dispatch.cpp:70]: I suppose this works, but we could also just call it directly?
`HWY_STATIC_DISPATCH(INT_SuperAbsolute)(args, dimensions, steps)`

**Review by jan-wassenberg — COMMENTED** [review]
Nice to see this simplified example :D

File: numpy/core/src/umath/absolute.dispatch.cpp (hunk 2 above)
Comment (Mousius) [reply numpy/core/src/umath/absolute.dispatch.cpp:37]: Oh, this is worse than that. It should be `LoadN` or `GatherLoadN`. I haven't written the proper load/store logic due to needing the whole set of `GatherLoadN`/`ScatterStoreN` as well.

File: numpy/core/src/umath/absolute.dispatch.cpp (hunk 1 above)
Comment (Mousius) [reply numpy/core/src/umath/absolute.dispatch.cpp:14]: Eek, I guess that's just a result size issue? I'll clean it up though.

File: numpy/core/src/umath/absolute.dispatch.cpp [hunk 4]
```diff
@@ -0,0 +1,129 @@
...
+      }
+    } else {
+      using TI = hwy::MakeSigned<T>;
+      const hn::Rebind<TI, hn::ScalableTag<T>> di;
+
+      const int lsize = sizeof(input_array[0]);
+      const npy_intp ssrc = steps[0] / lsize;
+      const npy_intp sdst = steps[1] / lsize;
+      auto load_index = hwy::AllocateAligned<TI>(hn::Lanes(d));
+      for (size_t i = 0; i < hn::Lanes(d); ++i) {
```
Comment (jan-wassenberg) [inline numpy/core/src/umath/absolute.dispatch.cpp:66]: We could actually vectorize this, just Mul() for <= 32 bit types, and perhaps converting to double and then Mul of that for 64-bit?

Comment (jan-wassenberg) [reply numpy/core/src/umath/absolute.dispatch.cpp:66]: Here's an example of how to specialize a generic function for certain types:
https://github.com/libjxl/libjxl/blob/e6202f7181eff36c78bfdb79aa9bd45c3d1d614b/lib/jxl/rational_polynomial-inl.h#L29

File: numpy/core/meson.build [hunk 5]
```diff
@@ -79,6 +79,10 @@ endif
 # important, because code in NumPy typically does not check the value but only
 # whether the symbol is defined. So `#define HAVE_SOMETHING 0` is wrong.
 
+#cmake = import('cmake')
+#hwy = cmake.subproject('highway')
```
Comment (rgommers) [inline numpy/core/meson.build:83]: @Mousius was this useful for local debugging? I'm working on Meson subproject usage elsewhere, so I'm curious about the details here.

**Comment by rgommers** [comment]
I tried to build this branch, to test with it a little bit. I had trouble checking out the submodule, since this points at a commit that's not part of the main repo. That can be updated now though, since the Gather/Scatter PR was merged.

A lot of CI jobs failed with:
```
/usr/bin/ld: build/temp.linux-x86_64-3.9/numpy/core/src/umath/umathmodule.o: relocation R_X86_64_PC32 against undefined hidden symbol `DOUBLE_absolute_SSE41' can not be used when making a shared object
```
and after merging `main` into this branch (which is needed to get SIMD support in the Meson build) I am also getting this error locally. @Mousius would you be able to rebase this? CI doesn't have to pass, but it'd be great if it worked with regular GCC on Linux x86-64. And don't worry about the `setup.py`-based build, since we're almost done with removing it (~4 CI jobs left to go).

**Comment by Mousius** [comment]
> I tried to build this branch, to test with it a little bit. I had trouble checking out the submodule, since this points at a commit that's not part of the main repo. That can be updated now though, since the Gather/Scatter PR was merged.
> 
> A lot of CI jobs failed with:
> 
> ```
> /usr/bin/ld: build/temp.linux-x86_64-3.9/numpy/core/src/umath/umathmodule.o: relocation R_X86_64_PC32 against undefined hidden symbol `DOUBLE_absolute_SSE41' can not be used when making a shared object
> ```
> 
> and after merging `main` into this branch (which is needed to get SIMD support in the Meson build) I am also getting this error locally. @Mousius would you be able to rebase this? CI doesn't have to pass, but it'd be great if it worked with regular GCC on Linux x86-64. And don't worry about the `setup.py`-based build, since we're almost done with removing it (~4 CI jobs left to go).

Ack. Apologies, I've been focussing on other things, I didn't originally do the spike on x86 so didn't pick this up, I'll take a look and follow up.

**Comment by Mousius** [comment]
@rgommers updated this, took me longer than I expected to unpick the new dispatch system. The issue was that I had not made sure all the values line up across the various dispatch files, so `meson.build` wasn't building that symbol.

Because I've taken examples from both `loops_autovec` and `loops_unary_fp` I had to align those with the example, but we can unpick that easily for other functions - if I built `SSE41` in addition it would use the same `HWY_NAMESPACE` as `AVX2` by accident. This is not an issue for a real world implementation.

**Comment by rgommers** [comment]
Thanks @Mousius, that works for me now. I'll try to play with this branch a bit.

**Pull request closed without merging**